Add SHACL ValidationListener

fkleedorfer opened the pull request:
Jira issue: https://issues.apache.org/jira/browse/JENA-2320

Adds a `ValidationListener` for SHACL validation that receives calls throughout the validation process.

namedgraph commented:
`ValidationListener` kind of asks for a more generic concept of `Listener` that could handle incoming requests.
Which would be like a [`ContainerRequestFilter`](https://docs.oracle.com/javaee/7/api/javax/ws/rs/container/ContainerRequestFilter.html) or [`MessageBodyReader`](https://docs.oracle.com/javaee/7/api/javax/ws/rs/ext/MessageBodyReader.html) in JAX-RS, but I'm not going to preach that again...

fkleedorfer commented:
> ValidationListener kind of asks for a more generic concept of Listener that could handle incoming requests.
Which would be like a [ContainerRequestFilter](https://docs.oracle.com/javaee/7/api/javax/ws/rs/container/ContainerRequestFilter.html) or [MessageBodyReader](https://docs.oracle.com/javaee/7/api/javax/ws/rs/ext/MessageBodyReader.html) in JAX-RS, but I'm not going to preach that again...

I am not sure I follow. If this is about the naming, we could rename the thing to `ValidationCallback` (however, I felt that a callback is something that has just one method, whereas a listener may have many). Otherwise, I don't really get it - what are the 'requests' you mention? The SHACL subsystem is quite low-level, depending only on the Graph API, and architecturally is nowhere near any HTTP interface, if I am not mistaken. Could you elaborate/clarify?

namedgraph commented:
@fkleedorfer my bad, I thought this was for Fuseki. I don't get how validation is invoked then...

fkleedorfer commented:
oh ok no problem. Maybe I should have been clearer in the PR description. It's a listener that allows to observe the SHACL validation process which is currently totally opaque apart from the ValidationReport that it produces.

namedgraph commented:
Basic use case in Fuseki would be validating incoming data against SHACL.

afs commented:
Looks good - I have tried it a little and got some basic printing working, nothing fancy but it is doing what I would expect

The only way I can think of reducing the number of listener calls is to have one "violated" method pass in the shape and use the fact hat shapes have visitors. But in doing that it is really only moving the code around. The form in the current PR is fine. Maybe with real use, we'll see a better way but at the moment it would be guessing (equivalently, premature optimization).

@fkleedorfer  If you want to make this "ready for review", it can be merged into the codebase and then will be in Jena 4.5.0. It can still change after that based on experience.

afs commented:
Looks OK, and, if I read it correctly, the default is null for the listener which means no calls.

fkleedorfer commented:
> Looks OK, and, if I read it correctly, the default is null for the listener which means no calls.

Correct, I tried to make the overhead minimal if no listener is used (event objects only generated if a listener is there).  However, the events do add some overhead through copying any Set<Node> that are passed to enforce immutability. Just passing the sets might be worth considering.

fkleedorfer commented:
> To get into Jena 4.5.0, please set this to "ready for review" from "draft"

Not really ready yet, here are a few things I'd still like to do:

* [ ] Lifecycle events for the whole validation process
* [ ] Improve naming of some event interfaces
* [ ] test this more with my use case (to make sure it does all I need - haven't really done that lately)

afs commented:
Hi there - Jena 4.5.0 is out and so we're at the beginning of the development cycle. A good time to merge new work! 

The maybe/hope is 4.6.0 is sooner than the usual 3-4 months because it will include the new in-memory standalone graph.

How are we on this PR?

fkleedorfer commented:
Haven't been able to do much lately, but:
* [x] Lifecycle events for the whole validation process
* [x] Improve naming of some event interfaces
* [ ] test this more with my use case (to make sure it does all I need - haven't really done that lately)
I looked into unit tests for this briefly and saw that for SHACL, there is a little framework for running the tests. I haven't found how to integrate tests for this here.

afs commented:
SHACL unit tests: I'm not sure the test manifest (which is from the WG test suite) is going to help test this. It tests the outcome of validation for the various constraints. It won't be a simple matter to add in a listener to test.

I think test for the listener will be java ones - setup a validation (that can be read from a file), then validate with listener that records whether it was called or not. Check it was called. It really doesn't have to be that much and (as before) the main thing is that the current use isn't affected and the existing test suite does that.

afs commented:
@fkleedorfer - How's the weather?! Just checking in to see where we are on this PR.

fkleedorfer commented:
> Merge now, and change is fine.

Ok.

> Do you have a specific list of changes in mind?

Two things in particular: 

## 1. More efficient collection handling in Events

Collections that are passed into Events' constructors are cloned and made unmodifiable in the constructor. This is wasteful, as we emit a lot of events, and it's quite possible nobody ever needs the collection. I'd like to change it so the reference to the original collection is stored in the event and a second reference is lazily initialized with the unmodifiable clone once the collection's getter is called. Alternatively, we could replace the reference to the original with the unmodifiable clone on first call to the getter. 

Example:

Current:
```
public abstract class AbstractConstraintEvaluationOnPathNodesEvent extends AbstractConstraintEvaluationForPathEvent
                implements
                ConstraintEvaluationOnPathNodesEvent {
    private final Set<Node> valueNodes;

    public AbstractConstraintEvaluationOnPathNodesEvent(ValidationContext vCxt, Shape shape, Node focusNode,
                    Constraint constraint, Path path, Set<Node> valueNodes) {
        super(vCxt, shape, focusNode, constraint, path);
        this.valueNodes = Collections.unmodifiableSet(new HashSet<>(valueNodes));
    }

    @Override public Set<Node> getValueNodes() {
        return valueNodes;
    }
}
```
Planned:
```
public abstract class AbstractConstraintEvaluationOnPathNodesEvent extends AbstractConstraintEvaluationForPathEvent
                implements
                ConstraintEvaluationOnPathNodesEvent {
    private final Set<Node> valueNodes;
    private Set<Node> valueNodesClone = null;

    public AbstractConstraintEvaluationOnPathNodesEvent(ValidationContext vCxt, Shape shape, Node focusNode,
                    Constraint constraint, Path path, Set<Node> valueNodes) {
        super(vCxt, shape, focusNode, constraint, path);
        this.valueNodes = valueNodes;
    }

    @Override public Set<Node> getValueNodes() {
        if (valueNodes == null) {
             // I think we don't need to synchronize here, not much harm will be done if called multiple times in parallel
             this.valueNodesClone = Collections.unmodifiableSet(new HashSet<>(this.valueNodes));
        }
        return valueNodesClone;
    }
}
```
## 2. Refactor the ValidationListener interface

I am not sure I like the ValidationListener, which has these methods:

```
public interface ValidationListener {

    void onConstraintViolatedEvent(ConstraintEvaluatedEvent e);

    void onConstraintSatisfiedEvent(ConstraintEvaluatedEvent e);

    void onValidationLifecycleEvent(ValidationLifecycleEvent e);

}
``` 

I don't like it much because it requires the following code in the ValidationContext:
```
            if (event instanceof ConstraintEvaluatedEvent){
                if (((ConstraintEvaluatedEvent)event).isValid()){
                    validationListener.onConstraintViolatedEvent((ConstraintEvaluatedEvent) event);
                } else {
                    validationListener.onConstraintSatisfiedEvent((ConstraintEvaluatedEvent) event);
                }
            } else if (event instanceof ValidationLifecycleEvent) {
                validationListener.onValidationLifecycleEvent((ValidationLifecycleEvent) event);
            } else {
                throw new IllegalArgumentException("Cannot handle event " + event);
            }
```
I.e., the interface is prejudiced about what will constitute a separation of events that is relevant to the client. (constraintViolated, constraintSatisfied, or lifecycle). I am not certain what would be better, but I am leaning toward just one method, `onValidationEvent`, and providing a utility class or a base class for listeners that make it easy to select the events one is interested in.

> Documentation will be needed sometime.

As an extension of https://jena.apache.org/documentation/shacl/index.html, I suppose?

fkleedorfer commented:
I'm ok with merging this PR and making the above changes at some later point in time

afs commented:
OK - do you want to quash the commits to remove the intermediate ones? I can do it if you prefer - it's more about best preserving your attribution.

fkleedorfer commented:
I made [the changes I wanted](https://github.com/apache/jena/pull/1256#issuecomment-1156415480) and squashed everything into one commit. Ready from my side.

fkleedorfer commented:
I have tested the PR with my client code a bit and I could add some minor tweaks and a generic configurable ValidationListener implementation. Should I adapt this PR or wait for the merge and make another?

afs commented:
Either works for me - 

I have a slight preference to get this merged because we're coming up to a release point (later this month if everything is done) it moves some work away from the time just before the release which is busy.

What would be good is:
* 2 or 3 sentences about the feature for the release announcement.
* Documentation to go [SHACL page](https://jena.apache.org/documentation/shacl/) ([jena website repo](https://github.com/apache/jena-site)). Users like examples!

fkleedorfer commented:
Regarding the documentation: the API is described at its 'facade', `ShaclValidator` - the ValidationListener does not appear in its interface, yet. Would you prefer the documentation to dig deeper and mention the `ValidationProc` or  `Vlib` or should we allow to pass the ValidationListener into the `ShaclValidator`'s methods?

afs commented:
If there is a mentioned (in text) on the [SHACL page](https://jena.apache.org/documentation/shacl/) that the feature exists it gives people a way to know the feature is there. A brief example is a bonus.

The javadoc can be changed to include the org.apache.jena.shacl.validation package if that helps.

Javadoc for the current version is published online at [SHACL javadoc](https://jena.apache.org/documentation/javadoc/shacl/org.apache.jena.shacl/org/apache/jena/shacl/package-summary.html).

fkleedorfer commented:
Unless there are change requests, I think this PR is done, here's a summary of the latest work:

* Added a Test class. It's more of a smoke test but it can easily be extended to make test fixtures for bugs or refactoring
* Related to the tests, added an `EventPredicates`, which has static methods for creating `Predicate<ValidationEvent>` instances, e.g. for checking if a `ConstraintEvaluatedEvent` reports a valid or an invalid check. Many more methods might be added to this as needed in the future.
* Added a configurable validation listener implementation, `HandlerBasedValidationListener`, making it easy to write such listeners
* Made a PR in the jena website repo describing the feature: https://github.com/apache/jena-site/pull/107
* added the `[...].shacl.validation` package (and `event` subpackge) to generated javadoc

afs commented:
Looks good except I get 48 warnings in Eclipse (I hope this isn't a mix up somewhere)

* Unused imports
* Missing `@Override`
* Unchecked casts

Could you please fix these up please?

For the 7 unchecked casts, please add a `@SupressWarning` with the least scope applicable (that might mean introducing and assigning to a local variable if necessary)? These mark "intentional" casts.

```
The import com.github.jsonldjava.utils.Obj is never used	AbstractValidationEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 21	Java Problem
The import java.util.Collections is never used	AbstractTargetShapesValidationEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 25	Java Problem
The import java.util.Collections is never used	ConstraintEvaluatedOnFocusNodeWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 26	Java Problem
The import java.util.Collections is never used	ConstraintEvaluatedOnPathNodesWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.Collections is never used	ConstraintEvaluatedOnSinglePathNodeWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.Collections is never used	ValueNodesDeterminedForPropertyShapeEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 26	Java Problem
The import java.util.HashSet is never used	AbstractTargetShapesValidationEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 26	Java Problem
The import java.util.HashSet is never used	ConstraintEvaluatedOnFocusNodeWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.HashSet is never used	ConstraintEvaluatedOnPathNodesWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 28	Java Problem
The import java.util.HashSet is never used	ConstraintEvaluatedOnSinglePathNodeWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 28	Java Problem
The import java.util.HashSet is never used	ValueNodesDeterminedForPropertyShapeEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.Objects is never used	DisjointConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 38	Java Problem
The import java.util.Objects is never used	AbstractTargetShapesValidationEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.Objects is never used	ConstraintEvaluatedOnFocusNodeEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.Objects is never used	ConstraintEvaluatedOnFocusNodeWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 28	Java Problem
The import java.util.Objects is never used	ConstraintEvaluatedOnFocusNodeWithSingleCompareNodeEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.Objects is never used	ConstraintEvaluatedOnPathNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.Objects is never used	ConstraintEvaluatedOnPathNodesWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 29	Java Problem
The import java.util.Objects is never used	ConstraintEvaluatedOnSinglePathNodeEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 28	Java Problem
The import java.util.Objects is never used	ConstraintEvaluatedOnSinglePathNodeWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 29	Java Problem
The import java.util.Objects is never used	ConstraintEvaluatedOnSinglePathNodeWithSingleCompareNodeEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 27	Java Problem
The import java.util.Objects is never used	ValueNodesDeterminedForPropertyShapeEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 28	Java Problem
The import java.util.Set is never used	DisjointConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 39	Java Problem
The import java.util.Set is never used	ValidationProc.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation	line 23	Java Problem
The import org.apache.jena.shacl.compact.writer.CompactOut.compactArrayNodes is never used	DisjointConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 41	Java Problem
The import org.apache.jena.shacl.lib.ShLib.displayStr is never used	DisjointConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 42	Java Problem
The import org.apache.jena.shacl.validation.event.AbstractConstraintEvaluationEvent is never used	LessThanOrEqualsConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 33	Java Problem
The import org.apache.jena.shacl.validation.event.AbstractTargetShapesValidationEvent is never used	TargetShapesValidationFinishedEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 23	Java Problem
The import org.apache.jena.shacl.validation.event.ConstraintEvaluatedOnFocusNodeWithCompareNodesEvent is never used	LessThanOrEqualsConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 34	Java Problem
The import org.apache.jena.shacl.validation.event.ConstraintEvaluatedOnPathNodesWithCompareNodesEvent is never used	DisjointConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 33	Java Problem
The import org.apache.jena.shacl.validation.event.ConstraintEvaluatedOnPathNodesWithCompareNodesEvent is never used	LessThanOrEqualsConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 35	Java Problem
The import org.apache.jena.shacl.validation.event.ConstraintEvaluatedOnSinglePathNodeWithCompareNodesEvent is never used	DisjointConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 34	Java Problem
The import org.apache.jena.shacl.validation.event.ConstraintEvaluatedOnSinglePathNodeWithCompareNodesEvent is never used	EqualsConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 27	Java Problem
The import org.apache.jena.shacl.validation.event.ConstraintEvaluatedOnSinglePathNodeWithSingleCompareNodeEvent is never used	LessThanConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 33	Java Problem
The import org.apache.jena.shacl.validation.event.ConstraintEvaluatedOnSinglePathNodeWithSingleCompareNodeEvent is never used	LessThanOrEqualsConstraint.java	/jena-shacl/src/main/java/org/apache/jena/shacl/engine/constraint	line 36	Java Problem
The import org.apache.jena.shacl.validation.event.ValidationLifecycleEvent is never used	TargetShapesValidationFinishedEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 24	Java Problem
The method getCompareNodes() of type ConstraintEvaluatedOnPathNodesWithCompareNodesEvent should be tagged with @Override since it actually overrides a superinterface method	ConstraintEvaluatedOnPathNodesWithCompareNodesEvent.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 47	Java Problem
The method handle(Consumer<T>) of type HandlerBasedValidationListener.HandlerBuilder<T> should be tagged with @Override since it actually overrides a superinterface method	HandlerBasedValidationListener.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation	line 218	Java Problem
The method iff(Predicate<ValidationEvent>) of type HandlerBasedValidationListener.HandlerBuilder<T> should be tagged with @Override since it actually overrides a superinterface method	HandlerBasedValidationListener.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation	line 214	Java Problem
The method toString() of type ImmutableLazyCollectionCopy<T> should be tagged with @Override since it actually overrides a superclass method	ImmutableLazyCollectionCopy.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 39	Java Problem
The method toString() of type ImmutableLazySetCopy<T> should be tagged with @Override since it actually overrides a superclass method	ImmutableLazySetCopy.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 39	Java Problem
Type safety: Unchecked cast from Class<? extends ValidationEvent>[] to Class<ValidationEvent>[]	HandlerBasedValidationListener.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation	line 152	Java Problem
Type safety: Unchecked cast from Class<capture#13-of ?> to Class<? extends ValidationEvent>	EventUtil.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 36	Java Problem
Type safety: Unchecked cast from Class<capture#3-of ?> to Class<? extends ValidationEvent>	EventUtil.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 31	Java Problem
Type safety: Unchecked cast from Consumer<capture#13-of ? super T> to Consumer<ValidationEvent>	HandlerBasedValidationListener.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation	line 82	Java Problem
Type safety: Unchecked cast from Consumer<T> to Consumer<ValidationEvent>	HandlerBasedValidationListener.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation	line 226	Java Problem
Type safety: Unchecked cast from HandlerBasedValidationListener.FilteredEventHandler to Consumer<T>	HandlerBasedValidationListener.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation	line 226	Java Problem
Type safety: Unchecked cast from ValidationEvent to E	EventPredicates.java	/jena-shacl/src/main/java/org/apache/jena/shacl/validation/event	line 81	Java Problem
```

fkleedorfer commented:
Will do, sorry.

fkleedorfer commented:
done

afs commented:
I'm still seeing 22 warnings. Did something not get pushed up?

Most I can fix.
The three I am less sure about are unchecked casts HandlerBasedValidationListener.java:

```
Type safety: Unchecked cast from Class<? extends ValidationEvent>[] to Class<ValidationEvent>[]  line 155
Type safety: Unchecked cast from Consumer<T> to Consumer<ValidationEvent> line 230
Type safety: Unchecked cast from HandlerBasedValidationListener.FilteredEventHandler to Consumer<T>   line 230
```
